Fixes #38 - Update overview & migrate 2 articles from source (#39)
MicrosoftDocs/PowerShell-Docs-Modules: reference/docs-conceptual/PSScriptAnalyzer/using-scriptanalyzer.md

ms.date: 03/22/2022
title: Using PSScriptAnalyzer
---
# Using PSScriptAnalyzer

This article describes various features of **PSScriptAnalyzer** and how to use them.

## Parser Errors

Beginning in version 1.18.0, **PSScriptAnalyzer** emits parser errors as diagnostic records in the
output stream.

```powershell
Invoke-ScriptAnalyzer -ScriptDefinition '"b" = "b"; function eliminate-file () { }'
```

```Output
RuleName            Severity   ScriptName Line Message
--------            --------   ---------- ---- -------
InvalidLeftHandSide ParseError            1    The assignment expression is not
                                               valid. The input to an
                                               assignment operator must be an
                                               object that is able to accept
                                               assignments, such as a variable
                                               or a property.
PSUseApprovedVerbs  Warning               1    The cmdlet 'eliminate-file' uses an
                                               unapproved verb.
```

> The RuleName is set to the `ErrorId` of the parser error.

Reviewer: ```suggestion
The RuleName is set to the **ErrorId** of the parser error.
```

the **categoryID** parameter to the name of the rule you want to suppress and set the **checkID**
parameter to a null or empty string. You can optionally add a third named parameter with a
justification for suppressing the message:

```powershell
function SuppressMe()
{
    [Diagnostics.CodeAnalysis.SuppressMessageAttribute('PSProvideCommentHelp', '', Justification='Just an example')]
    param()

    Write-Verbose -Message "I'm making a difference!"

}
```

Within the scope of the script, function, or parameter that you decorated, all rule violations are
suppressed.

To suppress a message on a specific parameter, set the **CheckId** parameter of the
`SuppressMessageAttribute` to the name of the parameter:

```powershell
function SuppressTwoVariables()
{
    [Diagnostics.CodeAnalysis.SuppressMessageAttribute('PSProvideDefaultParameterValue', 'b')]
    [Diagnostics.CodeAnalysis.SuppressMessageAttribute('PSProvideDefaultParameterValue', 'a')]
    param([string]$a, [int]$b)
    {
    }
}
```

Use the **Scope** property of the `SuppressMessageAttribute` to limit rule suppression to functions
or classes within the attribute's scope.

> Use the value `Function` to suppress violations on all functions within the attribute's scope. Use
> the value `Class` to suppress violations on all classes within the attribute's scope:

Reviewer: ```suggestion
Use the value **Function** to suppress violations on all functions within the attribute's scope. Use
the value **Class** to suppress violations on all classes within the attribute's scope:
```

```powershell
[Diagnostics.CodeAnalysis.SuppressMessageAttribute('PSProvideCommentHelp', '', Scope='Function')]
param(
)

function InternalFunction
{
    param()

    Write-Verbose -Message "I am invincible!"
}
```

You can further restrict suppression based on a function, parameter, class, variable or object's
name by setting the **Target** property of the `SuppressMessageAttribute` to a regular expression or
a wildcard pattern.

> For example, uppress `PSAvoidUsingWriteHost` rule violation in `start-bar` and `start-baz` but not

Reviewer: ```suggestion
For example, to suppress `PSAvoidUsingWriteHost` rule violation in `start-bar` and `start-baz` but not
```

in `start-foo` and `start-bam`:

```powershell
[System.Diagnostics.CodeAnalysis.SuppressMessageAttribute('PSAvoidUsingWriteHost', '', Scope='Function', Target='start-ba[rz]')]
param()
function start-foo {
    write-host "start-foo"
}

function start-bar {
    write-host "start-bar"
}

function start-baz {
    write-host "start-baz"
}

function start-bam {
    write-host "start-bam"
}
```

Suppress violations in all the functions:

```powershell
[Diagnostics.CodeAnalysis.SuppressMessageAttribute('PSAvoidUsingWriteHost', '', Scope='Function', Target='*')]
Param()
```

Suppress violation in `start-bar`, `start-baz` and `start-bam` but not in `start-foo`:

```powershell
[Diagnostics.CodeAnalysis.SuppressMessageAttribute('PSAvoidUsingWriteHost', '', Scope='Function', Target='start-b*')]
Param()
```

> **Note**: Parser Errors cannot be suppressed via the `SuppressMessageAttribute`

Reviewer: Should this be an MSDocs note block?

Author: Fixed

## Settings Support in ScriptAnalyzer

You can create settings that describe the ScriptAnalyzer rules to include or exclude based on

> `Severity`. Use the **Settings** parameter of `Invoke-ScriptAnalyzer` to specify configuration. This

Reviewer: ```suggestion
**Severity**. Use the **Settings** parameter of `Invoke-ScriptAnalyzer` to specify configuration. This
```

enables a user to create a custom configuration for a specific environment. We support the following
modes for specifying the settings file.

### Built-in Presets

ScriptAnalyzer ships a set of built-in presets that can be used to analyze scripts. For example, if
the user wants to run _PowerShell Gallery_ rules on their module, then they use the following
command.

```powershell
Invoke-ScriptAnalyzer -Path /path/to/module/ -Settings PSGallery -Recurse
```

> Along with `PSGallery` there are a few other built-in presets, including, `DSC` and
> `CodeFormatting`, that can be used. These presets can be tab completed for the **Settings** parameter.

Reviewer: ```suggestion
Along with **PSGallery** there are a few other built-in presets, including, **DSC** and
**CodeFormatting**, that can be used. These presets can be tab completed for the **Settings** parameter.
```

```json
{
    "powershell.scriptAnalysis.settingsPath": ".vscode/analyzersettings.psd1",
    "powershell.scriptAnalysis.enable": true,
}
```

## ScriptAnalyzer as a .NET library

ScriptAnalyzer engine and functionality can now be directly consumed as a library.

Here are the public interfaces:

```csharp
using Microsoft.Windows.PowerShell.ScriptAnalyzer;

public void Initialize(System.Management.Automation.Runspaces.Runspace runspace,
Microsoft.Windows.PowerShell.ScriptAnalyzer.IOutputWriter outputWriter,
[string[] customizedRulePath = null],
[string[] includeRuleNames = null],
[string[] excludeRuleNames = null],
[string[] severity = null],
[bool suppressedOnly = false],
[string profile = null])

public System.Collections.Generic.IEnumerable<DiagnosticRecord> AnalyzePath(string path,
    [bool searchRecursively = false])

public System.Collections.Generic.IEnumerable<IRule> GetRule(string[] moduleNames, string[] ruleNames)
```

## Violation Correction

Some violations can be fixed by replacing the violation causing content with a suggested
alternative. You can use the `-Fix` switch to automatically apply the suggestions. Since

> `Invoke-ScriptAnalyzer` implements `SupportsShouldProcess`, you can additionally use `-WhatIf` or

Reviewer: ```suggestion
`Invoke-ScriptAnalyzer` implements **SupportsShouldProcess**, you can additionally use `-WhatIf` or
```

`-Confirm` to find out which corrections would be applied. Be sure to use source control when

> applying those corrections since some changes, such as the one for `AvoidUsingPlainTextForPassword`,

Reviewer: ```suggestion
applying those corrections since some changes, such as the one for **AvoidUsingPlainTextForPassword**,
```

might require additional script modifications that cannot be made automatically. If your scripts are
sensitive to encoding, you should also check for that because the initial encoding can not be
preserved in all cases.

> The **SuggestedCorrections** property of the `ErrorRecord` enables quick-fix scenarios in editors

Reviewer: Nit: not sure about this
```suggestion
The **SuggestedCorrections** property of the error record enables quick-fix scenarios in editors
```

It is also possible to use the custom rules that are provided in the settings file in Visual Studio
Code. This is done by adding a Visual Studio Code workspace settings file (`.vscode/settings.json`).

```json
{
    "powershell.scriptAnalysis.settingsPath": ".vscode/analyzersettings.psd1",
    "powershell.scriptAnalysis.enable": true,
}
```

## ScriptAnalyzer as a .NET library

ScriptAnalyzer engine and functionality can now be directly consumed as a library.

Here are the public interfaces:

```csharp
using Microsoft.Windows.PowerShell.ScriptAnalyzer;

public void Initialize(System.Management.Automation.Runspaces.Runspace runspace,
Microsoft.Windows.PowerShell.ScriptAnalyzer.IOutputWriter outputWriter,
[string[] customizedRulePath = null],
[string[] includeRuleNames = null],
[string[] excludeRuleNames = null],
[string[] severity = null],
[bool suppressedOnly = false],
[string profile = null])

public System.Collections.Generic.IEnumerable<DiagnosticRecord> AnalyzePath(string path,
    [bool searchRecursively = false])

public System.Collections.Generic.IEnumerable<IRule> GetRule(string[] moduleNames, string[] ruleNames)
```

## Violation Correction

> Some violations can be fixed by replacing the violation causing content with a suggested
> alternative. You can use the `-Fix` switch to automatically apply the suggestions. Since
> `Invoke-ScriptAnalyzer` implements **SupportsShouldProcess**, you can additionally use `-WhatIf` or
> `-Confirm` to find out which corrections would be applied. Be sure to use source control when
> applying those corrections since some changes, such as the one for **AvoidUsingPlainTextForPassword**,
> might require additional script modifications that cannot be made automatically. If your scripts are
> sensitive to encoding, you should also check for that because the initial encoding can not be
> preserved in all cases.

Reviewer: ```suggestion
You can use the **Fix** switch to to automatically replace violation-causing content with a
suggested alternative. Additionally, because `Invoke-ScriptAnalyzer` implements
**SupportsShouldProcess**, you can use **WhatIf** or **Confirm** to find out which corrections
would be applied. You should use source control when applying corrections as some changes, such as
the one for **AvoidUsingPlainTextForPassword**, might require additional script modifications that
can't be made automatically. Because initial encoding can't always be preserved when you
automatically apply suggestions, you should check your file's encoding if your scripts depend on a
particular encoding.
```

The **SuggestedCorrections** property of the error record enables quick-fix scenarios in editors
like VSCode. We provide valid **SuggestedCorrection**  for the following rules:

> - AvoidAlias.cs
> - AvoidUsingPlainTextForPassword.cs
> - MisleadingBacktick.cs
> - MissingModuleManifestField.cs
> - UseToExportFieldsInManifest.cs

Reviewer: ```suggestion
- **AvoidAlias**
- **AvoidUsingPlainTextForPassword**
- **MisleadingBacktick**
- **MissingModuleManifestField**
- **UseToExportFieldsInManifest**
```
